signer/xpi/xpi.go

```diff
@@ -556,16 +557,70 @@ func VerifySignedFile(signedFile signer.SignedFile, truststore *x509.CertPool, o
 		return errors.Wrap(err, "xpi: error verifying PKCS7 signature for signed file")
 	}
 
-	if len(opts.COSEAlgorithms) > 0 {
-		err = verifyCOSEManifest(signedFile)
-		if err != nil {
-			return errors.Wrap(err, "xpi: error verifying COSE manifest for signed file")
-		}
-		err = verifyCOSESignatures(signedFile, truststore, opts)
-		if err != nil {
-			return errors.Wrap(err, "xpi: error verifying COSE signatures for signed file")
+	coseSig, err := readFileFromZIP(signedFile, coseSigPath)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "xpi: no COSE signature found, skipping")
+		return nil
+	}
+	extId, err := GetXpiID(signedFile)
+	if err != nil {
+		return errors.Wrap(err, "xpi: failed to get extension id from signed file")
+	}
+	opts := Options{ID: extId}
+	msg, err := cose.Unmarshal(coseSig)
+	if err != nil {
+		return errors.Wrap(err, "xpi: error reading COSE signature")
+	}
+	for _, sig := range msg.(cose.SignMessage).Signatures {
+		if _, ok := sig.Headers.Protected[algHeaderValue]; ok {
+			var alg *cose.Algorithm
+			algValue, ok := sig.Headers.Protected[algHeaderValue]
+			if !ok {
+				return errors.Errorf("xpi: missing expected alg in Protected Headers")
+			}
+			if algInt, ok := algValue.(int); ok {
+				alg = intToCOSEAlg(algInt)
+			}
+			if alg == nil {
+				return errors.Errorf("xpi: alg %v is not supported", algValue)
+			}
+			opts.COSEAlgorithms = append(opts.COSEAlgorithms, alg.Name)
 		}
 	}
+	if len(opts.COSEAlgorithms) < 1 {
+		return errors.Errorf("xpi: COSE signature exists but no valid algorithm could be found")
+	}
+	err = verifyCOSEManifest(signedFile)
+	if err != nil {
+		return errors.Wrap(err, "xpi: error verifying COSE manifest for signed file")
+	}
+	err = verifyCOSESignatures(signedFile, truststore, opts)
+	if err != nil {
+		return errors.Wrap(err, "xpi: error verifying COSE signatures for signed file")
+	}
 
 	return nil
 }
+
+// XpiManifest represent parts of the structure of an extension manifest
+type XpiManifest struct {
+	BrowserSpecificSettings struct {
+		Gecko struct {
+			ID string `json:"id"`
+		} `json:"gecko"`
+	} `json:"browser_specific_settings"`
+}
+
+// GetXpiID returns the string ID of an extension from its manifest
+func GetXpiID(xpiFile signer.SignedFile) (string, error) {
```

Comment:
+1 might be worth validating the ID = OU in the PK7 sig too (we can open an issue for that)

Reply:
Yep. It could be done by diving in `xpi.Signature.p7.Certificates`, but that's a bit annoying (have to parse signerInfo to find the right cert) and I'm not sure it's worth it since we want to kill pkcs7 down the road.